internal/cas/tree.go

```diff
@@ -99,40 +101,139 @@ func ParseTree(output, path string) (*Tree, error) {
 func (t *Tree) LinkTree(ctx context.Context, store *Store, targetDir string) error {
 	content := NewContent(store)
 
-	for _, entry := range t.entries {
-		// Check for cancellation
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		default:
-		}
+	dirsToCreate := make(map[string]struct{}, len(t.entries))
+
+	type workItem struct {
+		itemType string
+		entry    TreeEntry
+		path     string
+		dirPath  string
+	}
 
+	workItems := make([]workItem, 0, len(t.entries))
+
+	for _, entry := range t.entries {
 		entryPath := filepath.Join(targetDir, entry.Path)
-		if err := os.MkdirAll(filepath.Dir(entryPath), DefaultDirPerms); err != nil {
-			return wrapError("mkdir_all", entryPath, err)
-		}
+		dirPath := filepath.Dir(entryPath)
+
+		dirsToCreate[dirPath] = struct{}{}
 
+		// If the parent directory is in dirsToCreate,
+		// we can remove it, since it will be created
+		// when creating the subtree anyways.
+		parentDirPath := filepath.Dir(dirPath)
+		delete(dirsToCreate, parentDirPath)
+
+		// Create work items based on entry type
 		switch entry.Type {
 		case "blob":
-			if err := content.Link(entry.Hash, entryPath); err != nil {
-				return wrapError("link_blob", entryPath, err)
-			}
+			workItems = append(workItems, workItem{
+				itemType: "link",
+				entry:    entry,
+				path:     entryPath,
+				dirPath:  dirPath,
+			})
 		case "tree":
-			treeData, err := content.Read(entry.Hash)
-			if err != nil {
-				return wrapError("read_tree", entry.Hash, err)
-			}
+			workItems = append(workItems, workItem{
+				itemType: "subtree",
+				entry:    entry,
+				path:     entryPath,
+				dirPath:  dirPath,
+			})
+		}
+	}
 
-			subTree, err := ParseTree(string(treeData), entryPath)
-			if err != nil {
-				return wrapError("parse_tree", entry.Hash, err)
-			}
+	for dirPath := range dirsToCreate {
+		if err := os.MkdirAll(dirPath, DefaultDirPerms); err != nil {
+			return wrapError("mkdir_all", dirPath, err)
+		}
+	}
 
-			if err := subTree.LinkTree(ctx, store, entryPath); err != nil {
-				return wrapError("link_subtree", entryPath, err)
+	scalingFactor := 2
+	maxWorkers := max(1, runtime.NumCPU()/scalingFactor)
+	workChan := make(chan workItem, len(workItems))
+	errChan := make(chan error, 1)
+
+	var wg sync.WaitGroup
+
+	for i := 0; i < maxWorkers; i++ {
```

Comment:
Maybe consider using an [errgroup](https://pkg.go.dev/golang.org/x/sync/errgroup) here?  
It looks like we're only returning a single error, so we might not be capturing all errors from the workers.